File: src/runner/preflight.ts
import type { SessionsConfig } from "../config";
import type { InvocationContext } from "../context";
import type { LoadedWorkflow } from "../workflows/types";
import type { RunnerDeps } from "./types";

export type Preflight =
  | { ok: true; session: string; sessionFailure?: string; agent: string }
  | { ok: false; error: string };

/** Resolve {session} and {agent} preconditions; session extraction failure is non-fatal. */
export async function resolvePreflight(
  workflow: LoadedWorkflow,
  ctx: InvocationContext,
  agents: Record<string, string[]>,
  sessions: SessionsConfig,
  deps: RunnerDeps,
): Promise<Preflight> {
  let session = "";
  // Extraction failure aborts before steps and does not trigger on_error.
  let sessionFailure: string | undefined;
  if (workflow.needsSession) {
    if (!ctx.paneId) {
      return { ok: false, error: "session handoff must be launched from an agent pane" };
    }
    try {
      session = await deps.sessionText(ctx.paneId, sessions);
    } catch (err) {
      sessionFailure = err instanceof Error ? err.message : String(err);
    }
  }

  let agent = "";
  if (workflow.needsInvokingAgent) {
    if (!ctx.paneId) {
      return { ok: false, error: "invoking agent unresolved — run from agent pane" };
    }
    try {
      const label = await deps.agentLabel(ctx.paneId);
      if (!agents[label]) {
        return {
          ok: false,
          error: `invoking agent '${label}' not in config — add it under agents:`,
        };
      }
      agent = label;
    } catch (err) {
      return {
        ok: false,
        error:
          err instanceof Error ? err.message : "invoking agent unresolved — run from agent pane",
      };
    }
  }
  return { ok: true, session, sessionFailure, agent };
}

File: rust/src/runner/preflight.rs
//! Pre-run precondition resolution. Port of `src/runner/preflight.ts`:
//! `{session}` and `{agent}` requirements; session extraction failure is
//! non-fatal (recovery can fall back to `{pane}`).

use super::deps::Herdr;
use crate::config::{AgentsConfig, SessionsConfig};
use crate::runner::context::InvocationContext;
use crate::workflow::load::LoadedWorkflow;

/// `Preflight`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Preflight {
    /// Preconditions satisfied; `session_failure` still runs `on_fail`.
    Ok {
        session: String,
        session_failure: Option<String>,
        agent: String,
    },
    /// Hard failure before any step runs.
    Err(String),
}

/// `resolvePreflight`.
pub fn resolve_preflight(
    workflow: &LoadedWorkflow,
    ctx: &InvocationContext,
    agents: &AgentsConfig,
    sessions: &SessionsConfig,
    herdr: &dyn Herdr,
) -> Preflight {
    let mut session = String::new();
    let mut session_failure: Option<String> = None;
    if workflow.needs_session {
        let Some(pane_id) = ctx.pane_id.as_deref() else {
            return Preflight::Err(
                "session handoff must be launched from an agent pane".to_string(),
            );
        };
        match herdr.session_text(pane_id, sessions) {
            Ok(text) => session = text,
            Err(error) => session_failure = Some(error.message),
        }
    }

    let mut agent = String::new();
    if workflow.needs_invoking_agent {
        let Some(pane_id) = ctx.pane_id.as_deref() else {
            return Preflight::Err("invoking agent unresolved — run from agent pane".to_string());
        };
        match herdr.agent_label(pane_id) {
            Ok(label) => {
                if !agents.contains_key(&label) {
                    return Preflight::Err(format!(
                        "invoking agent '{label}' not in config — add it under agents:"
                    ));
                }
                agent = label;
            }
            Err(error) => return Preflight::Err(error.message),
        }
    }
    Preflight::Ok {
        session,
        session_failure,
        agent,
    }
}
